sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,807 @@
+        centroids_half_l2_norm,            # IN             (n_clusters,)
+        per_sample_pseudo_inertia,         # OUT            (n_samples,)
+        new_centroids_t_private_copies,    # OUT            (n_private_copies, n_features, n_clusters)
+        cluster_sizes_private_copies,      # OUT            (n_private_copies, n_clusters)
+    ):
+    # fmt: on
+        """One full iteration of LLoyd's k-means.
+
+        The kernel is meant to be invoked on a 1D grid spanning the data samples of
+        the training set in parallel.
+
+        Each work-item will assign that sample to its nearest centroid and accumulate
+        the feature values of the data samples into the new version of the centroid
+        for the next iteration. The centroid assignment and centroid update steps are
+        performed in a single fused kernel to avoid introducing a large intermediate
+        label assignment array to be re-read from global memory before performing the
+        centroid update step.
+
+        To avoid atomic contention when performing the centroid updates concurrently
+        for different data samples, we use private copies of the new centroid array in
+        global memory. Those copies are meant to be re-reduced afterwards.
+
+        The distance is the euclidean distance. Note that it is not necessary to
+        compute the exact value to find the closest centroid. Indeed, minimizing
+            |x-c|^2 = |x|^2 - 2<x.c> + |c|^2
+        over c for a given x amounts to minimizing
+            (1/2)c^2 - xc .
+        Moreover the value (1/2)c^2 has been pre-computed in the array
+        centroids_half_l2_norm to reduce the overall number of floating point
+        operations in the kernel.
+        """
+        sample_idx = dpex.get_global_id(0)
+        local_work_id = dpex.get_local_id(0)
+
+        # This array in shared memory is used as a sliding array over values of
+        # current_centroids_t. During each iteration in the inner loop, a new one is
+        # loaded and used by all work items in the work group to compute partial
+        # results. The array slides over the features in the outer loop, and over the
+        # samples in the inner loop.
+        centroids_window = dpex.local.array(shape=centroids_window_shape, dtype=dtype)
+
+        # This array in shared memory is used as a sliding array over the centroids.
+        # It contains values of centroids_half_l2_norm for each centroid in the sliding
+        # centroids_window array. It is updated once per iteration in the outer loop.
+        centroids_window_half_l2_norm = dpex.local.array(
+            shape=window_n_centroids, dtype=dtype
+        )
+
+        # In the inner loop each work item accumulates in private memory the
+        # dot product of the sample at the sample_idx relatively to each centroid
+        # in the window.
+        partial_dot_products = dpex.private.array(shape=window_n_centroids, dtype=dtype)
```

Comment:
```suggestion
        dot_products = dpex.private.array(shape=window_n_centroids, dtype=dtype)
```

Reply:
The name "partial" is used in other kernels to call something that is going to be accumulated, so it's only a part of the expected quantity as long as the accumulation is not over, is it what you understood too ? If this name needs changing other similar variable names must be updated accordingly.

Reply:
Ah, I see looking at other naming using this suffix, e.g. `partial_l2_norm`. I think using this prefix makes explicit that this is going to be computing iteratively, so do ignore this suggestion (I am going to edit or remove the other similar suggestions).

Also, I like the `pseudo_` prefix as well because it makes it explicit that we are using surrogates (e.g. we are not using the full inertia).